### PR TITLE
DEPR/API: Non-ns precision in Index constructors

### LINKED ISSUE #24739: 0.24.0rc1: infer_dtype(DatetimeIndex) returns "datetime64" not "datetime64[ns]"
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

datetime_index = pd.DatetimeIndex(['2017-08-01', '2017-08-02'])
print(datetime_index)
# DatetimeIndex(['2017-08-01', '2017-08-02'], dtype='datetime64[ns]', freq=None)
print(datetime_index.dtype)
# datetime64[ns]
infered_dtype = pd.api.types.infer_dtype(datetime_index, skipna=True)
print(infered_dtype)
# datetime64
print(pd.Index(['2017-08-01', '2017-08-02'], dtype=infered_dtype))
# Traceback (most recent call last):
#   File "/tmp/a.py", line 8, in <module>
#     print(pd.Index(['2017-08-01', '2017-08-02'], dtype=infered_dtype))
#   File "/usr/local/lib/python3.7/dist-packages/pandas/core/indexes/base.py", line 308, in __new__
#     dtype=dtype, **kwargs)
#   File "/usr/local/lib/python3.7/dist-packages/pandas/core/indexes/datetimes.py", line 303, in __new__
#     int_as_wall_time=True)
#   File "/usr/local/lib/python3.7/dist-packages/pandas/core/arrays/datetimes.py", line 368, in _from_sequence
#     ambiguous=ambiguous, int_as_wall_time=int_as_wall_time)
#   File "/usr/local/lib/python3.7/dist-packages/pandas/core/arrays/datetimes.py", line 1706, in sequence_to_dt64ns
#     dtype = _validate_dt64_dtype(dtype)
#   File "/usr/local/lib/python3.7/dist-packages/pandas/core/arrays/datetimes.py", line 1993, in _validate_dt64_dtype
#     .format(dtype=dtype))
# ValueError: Unexpected value for 'dtype': 'datetime64'. Must be 'datetime64[ns]' or DatetimeTZDtype'.
```

#### Problem description

We can't use dtype inferred from `DatetimeIndex` to convert an `Index` to `DatetimeIndex` with pandas 0.24.0rc1.

[pyarrow](https://github.com/apache/arrow/tree/master/python) uses this logic to convert Arrow objects to pandas objects.

FYI: Here are related codes:

  * https://github.com/apache/arrow/blob/master/python/pyarrow/pandas_compat.py#L36
  * https://github.com/apache/arrow/blob/master/python/pyarrow/pandas_compat.py#L114
  * https://github.com/apache/arrow/blob/master/python/pyarrow/pandas_compat.py#L235
  * https://github.com/apache/arrow/blob/master/python/pyarrow/pandas_compat.py#L789

FYI: Here is an error in pyarrow test: https://travis-ci.org/kszucs/crossbow/builds/478558634#L2724-L2788

#24478 introduces a validation for `datetime64`. But `pd.api.types.infer_dtype` still returns `'datetime64'` for `DatetimeIndex`.

The following change fixes this problem. But I'm not sure whether this is a regression or pyarrow's use case is wrong.

```diff
diff --git a/pandas/_libs/lib.pyx b/pandas/_libs/lib.pyx
index 85eb6c342..6271d1204 100644
--- a/pandas/_libs/lib.pyx
+++ b/pandas/_libs/lib.pyx
@@ -928,7 +928,7 @@ _TYPE_MAP = {
     'U': 'unicode' if PY2 else 'string',
     'bool': 'boolean',
     'b': 'boolean',
-    'datetime64[ns]': 'datetime64',
+    'datetime64[ns]': 'datetime64[ns]',
     'M': 'datetime64',
     'timedelta64[ns]': 'timedelta64',
     'm': 'timedelta64',
```

#### Expected Output

```python
import pandas as pd

datetime_index = pd.DatetimeIndex(['2017-08-01', '2017-08-02'])
print(datetime_index)
# DatetimeIndex(['2017-08-01', '2017-08-02'], dtype='datetime64[ns]', freq=None)
print(datetime_index.dtype)
# datetime64[ns]
infered_dtype = pd.api.types.infer_dtype(datetime_index, skipna=True)
print(infered_dtype)
# datetime64[ns]
print(pd.Index(['2017-08-01', '2017-08-02'], dtype=infered_dtype))
# DatetimeIndex(['2017-08-01', '2017-08-02'], dtype='datetime64[ns]', freq=None)
```

#### Output of ``pd.show_versions()``

<details>

```text
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.19.0-1-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: ja_JP.UTF-8
LOCALE: ja_JP.UTF-8

pandas: 0.24.0rc1
pytest: 3.10.1
pip: 18.1
setuptools: 40.6.2
Cython: 0.29.2
numpy: 1.16.0rc2
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.9
blosc: None
bottleneck: None
tables: 3.4.4
numexpr: 2.6.9
feather: None
matplotlib: 3.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
```

</details>

### LINKED ISSUE #24753: [0.24.0rc1] passing dtype='M8' to Index raises
In 0.23.4

```python
In [1]: import pandas as pd

In [2]: pd.Index([], dtype='M8')
Out[2]: DatetimeIndex([], dtype='datetime64[ns]', freq=None)
```

In 0.24.0rc1

```python
In [2]: pd.Index([], dtype='M8')
```
```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-09844a8ae29c> in <module>
----> 1 pd.Index([], dtype='M8')

~/Envs/dask-dev/lib/python3.7/site-packages/pandas/core/indexes/base.py in __new__(cls, data, dtype, copy, name, fastpath, tupleize_cols, **kwargs)
    306             else:
    307                 result = DatetimeIndex(data, copy=copy, name=name,
--> 308                                        dtype=dtype, **kwargs)
    309                 return result
    310

~/Envs/dask-dev/lib/python3.7/site-packages/pandas/core/indexes/datetimes.py in __new__(cls, data, freq, start, end, periods, tz, normalize, closed, ambiguous, dayfirst, yearfirst, dtype, copy, name, verify_integrity)
    301             data, dtype=dtype, copy=copy, tz=tz, freq=freq,
    302             dayfirst=dayfirst, yearfirst=yearfirst, ambiguous=ambiguous,
--> 303             int_as_wall_time=True)
    304
    305         subarr = cls._simple_new(dtarr, name=name,

~/Envs/dask-dev/lib/python3.7/site-packages/pandas/core/arrays/datetimes.py in _from_sequence(cls, data, dtype, copy, tz, freq, dayfirst, yearfirst, ambiguous, int_as_wall_time)
    366             data, dtype=dtype, copy=copy, tz=tz,
    367             dayfirst=dayfirst, yearfirst=yearfirst,
--> 368             ambiguous=ambiguous, int_as_wall_time=int_as_wall_time)
    369
    370         freq, freq_infer = dtl.validate_inferred_freq(freq, inferred_freq,

~/Envs/dask-dev/lib/python3.7/site-packages/pandas/core/arrays/datetimes.py in sequence_to_dt64ns(data, dtype, copy, tz, dayfirst, yearfirst, ambiguous, int_as_wall_time)
   1704     inferred_freq = None
   1705
-> 1706     dtype = _validate_dt64_dtype(dtype)
   1707
   1708     if not hasattr(data, "dtype"):

~/Envs/dask-dev/lib/python3.7/site-packages/pandas/core/arrays/datetimes.py in _validate_dt64_dtype(dtype)
   1991             raise ValueError("Unexpected value for 'dtype': '{dtype}'. "
   1992                              "Must be 'datetime64[ns]' or DatetimeTZDtype'."
-> 1993                              .format(dtype=dtype))
   1994     return dtype
   1995

ValueError: Unexpected value for 'dtype': 'datetime64'. Must be 'datetime64[ns]' or DatetimeTZDtype'.
```

We want that to  raise eventually; `ns` precision should be specified. But we should maybe deprecate the old behavior first?

The best place to do it is probably before we get to `arrays`, so in `DatetimeIndex.__new__` we can check for `M8` specifically, warn, then pass through `M8[ns]`.

We should also check

- `np.dtype("M8")`
- `'m8'`
- `np.dtype('m8')`

---

It's less clear what we should do for something like `M8[us]`. In the past, we used to ignore the precision

```python
In [15]: pd.Index(list(pd.date_range('2000', periods=4).asi8), dtype='M8[us]')
Out[15]: DatetimeIndex(['2000-01-01', '2000-01-02', '2000-01-03', '2000-01-04'], dtype='datetime64[ns]', freq=None)
```

this should arguably raise now...

### PR DESCRIPTION
**Pull request by TomAugspurger**
This deprecates passing dtypes without a precision to DatetimeIndex
and TimedeltaIndex

```python
In [2]: pd.DatetimeIndex(['2000'], dtype='datetime64')
/Users/taugspurger/.virtualenvs/pandas-dev/bin/ipython:1: FutureWarning: Passing in 'datetime64' dtype with no precision is deprecated
and will raise in a future version. Please pass in
'datetime64[ns]' instead.
  #!/Users/taugspurger/Envs/pandas-dev/bin/python3
Out[2]: DatetimeIndex(['2000-01-01'], dtype='datetime64[ns]', freq=None)
```

Closes https://github.com/pandas-dev/pandas/issues/24739
Closes https://github.com/pandas-dev/pandas/issues/24753

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.rst [hunk 1]
```diff
@@ -436,6 +436,7 @@ Backwards incompatible API changes
 - Incorrectly passing a :class:`DatetimeIndex` to :meth:`MultiIndex.from_tuples`, rather than a sequence of tuples, now raises a ``TypeError`` rather than a ``ValueError`` (:issue:`24024`)
 - :func:`pd.offsets.generate_range` argument ``time_rule`` has been removed; use ``offset`` instead (:issue:`24157`)
 - In 0.23.x, pandas would raise a ``ValueError`` on a merge of a numeric column (e.g. ``int`` dtyped column) and an ``object`` dtyped column (:issue:`9780`). We have re-enabled the ability to merge ``object`` and other dtypes; pandas will still raise on a merge between a numeric and an ``object`` dtyped column that is composed only of strings (:issue:`21681`)
+- :class:`DatetimeIndex` and :class:`TimedeltaIndex` no longer ignore the dtype precision. Passing a non-nanosecond resolution dtype will raise a ``ValueError`` (:issue:`24753`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.rst]: What do people think about this? In 0.23.4 we ignore the dtype precision

```python
In [3]: pd.DatetimeIndex(['2000'], dtype='datetime64[us]')
Out[3]: DatetimeIndex(['2000-01-01'], dtype='datetime64[ns]', freq=None)
```

Should that raise now? Or should we deprecate first?

File: pandas/core/arrays/timedeltas.py [hunk 2]
```diff
@@ -997,6 +989,30 @@ def objects_to_td64ns(data, unit="ns", errors="raise"):
...
+                'timedelta64[ns]' instead.""")
+            warnings.warn(msg, FutureWarning, stacklevel=4)
+    except TypeError:
+        # extension dtype
+        pass
+
+    try:
+        dtype_mismatch = dtype != _TD_DTYPE
+    except TypeError:
+        raise ValueError(_BAD_DTYPE.format(dtype=dtype))
```
Comment (TomAugspurger) [inline pandas/core/arrays/timedeltas.py]: Note the change from TypeError to ValueError, when compared with https://github.com/pandas-dev/pandas/pull/24806/files#diff-57368e94ebd4c37c96935ab9862e3bc7L169. ValueError is more consistent with DatetimeArray.

File: doc/source/whatsnew/v0.24.0.rst (hunk 1 above)
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.rst]: we actually can convert this (and I *think* we do?)

Comment (jreback) [reply doc/source/whatsnew/v0.24.0.rst]: though I would not be averse to deprecating for consistency reasons

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.rst (hunk 1 above)
Comment (TomAugspurger) [reply doc/source/whatsnew/v0.24.0.rst]: We do convert an *array* with non-ns precision dtype. That's similar to converting any sequence of datetime-like things

```python
In [7]: arr = np.array([1, 2], dtype='datetime64[us]')

In [8]: pd.DatetimeIndex(arr)
Out[8]: DatetimeIndex(['1970-01-01 00:00:00.000001', '1970-01-01 00:00:00.000002'], dtype='datetime64[ns]', freq=None)
```

The issue is when the user says `DatetimeIndex(..., dtype='datetime64[us]')`. That's like saying "I want the dtype of the *output* DatetimeIndex to be datetime64[us]", which isn't supported.

**Comment by TomAugspurger** [comment]
Fixed the merge conflict.

I'm inclined to keep the change of `DatetimeIndex([1, 2], dtype='datetime64[us])`, as the user is unambiguously asking for something wrong (in comparison with `'datetime64'`, which is ambiguous, so we deprecate it).

**Comment by jbrockmendel** [comment]
No strong opinion on raise vs warn for datetime64[us]

File: doc/source/whatsnew/v0.24.0.rst [hunk 3]
```diff
@@ -1259,6 +1260,7 @@ Deprecations
 - :meth:`Series.nonzero` is deprecated and will be removed in a future version (:issue:`18262`)
 - Passing an integer to :meth:`Series.fillna` and :meth:`DataFrame.fillna` with ``timedelta64[ns]`` dtypes is deprecated, will raise ``TypeError`` in a future version.  Use ``obj.fillna(pd.Timedelta(...))`` instead (:issue:`24694`)
 - ``Series.cat.categorical``, ``Series.cat.name`` and ``Sersies.cat.index`` have been deprecated. Use the attributes on ``Series.cat`` or ``Series`` directly. (:issue:`24751`).
+- Passing a dtype without a precision like ``np.dtype('datetime64')`` or ``timedelta64`` to :class:`DatetimeIndex` and :class:`TimedeltaIndex` is now deprecated. Use the nanosecond-precision dtype instead (:issue:`24753`).
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.24.0.rst]: Also to `Index` ?

**Review by jorisvandenbossche — APPROVED** [review]
Fine with this approach

**Comment by jorisvandenbossche** [comment]
> On 0.23 that silently ignored the precision.

Pedantic note: it doesn't really "ignores the precision" in this example case, as it is doing "datetime string parsing", even if you pass `dtype=datetime64[ns]` it doesn't use the ns precision to interpret the string value '2000'. 
But we do ignore it for integer input:

```
In [10]: pd.DatetimeIndex([2000], dtype='datetime64[s]')
Out[10]: DatetimeIndex(['1970-01-01 00:00:00.000002'], dtype='datetime64[ns]', freq=None)
```

so in the end my note doesn't matter :), because here it is indeed being ignored, and fine to raise an error here.

**Comment by jorisvandenbossche** [comment]
> fine to raise an error here.

Actually, this already seems to be failing on master:

```
In [86]: pd.DatetimeIndex([2000], dtype='datetime64[s]')                                                                                                                        
...
ValueError: Unexpected value for 'dtype': 'datetime64[s]'. Must be 'datetime64[ns]' or DatetimeTZDtype'.
```

So we already changed this (by accident?) compared to 0.23.4.

File: doc/source/whatsnew/v0.24.0.rst (hunk 3 above)
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.rst]: I think we raise on this to Series now, so this is just conforming DTI?

File: pandas/core/arrays/timedeltas.py [hunk 4]
```diff
@@ -997,6 +989,30 @@ def objects_to_td64ns(data, unit="ns", errors="raise"):
     return result.view('timedelta64[ns]')
 
 
+def _validate_td64_dtype(dtype):
+    try:
+        if dtype == np.dtype("timedelta64"):
+            dtype = _TD_DTYPE
+            msg = textwrap.dedent("""\
+                Passing in 'timedelta' dtype with no precision is deprecated
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: why are you warning? this *already* raises
```
In [5]: pd.TimedeltaIndex([1,2,3],dtype='timedelta64')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-65a1792bd679> in <module>()
----> 1 pd.TimedeltaIndex([1,2,3],dtype='timedelta64')

~/pandas-dev/pandas/core/indexes/timedeltas.py in __new__(cls, data, unit, freq, start, end, periods, closed, dtype, copy, name, verify_integrity)
    224 
    225         tdarr = TimedeltaArray._from_sequence(data, freq=freq, unit=unit,
--> 226                                               dtype=dtype, copy=copy)
    227         return cls._simple_new(tdarr._data, freq=tdarr.freq, name=name)
    228 

~/pandas-dev/pandas/core/arrays/timedeltas.py in _from_sequence(cls, data, dtype, copy, freq, unit)
    206                        freq=None, unit=None):
    207         if dtype != _TD_DTYPE:
--> 208             raise ValueError("Only timedelta64[ns] dtype is valid.")
    209 
    210         freq, freq_infer = dtl.maybe_infer_freq(freq)

ValueError: Only timedelta64[ns] dtype is valid.
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/timedeltas.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/arrays/timedeltas.py]: Did you see this issue? This didn't raise on 0.23.4.

**Comment by TomAugspurger** [comment]
@jorisvandenbossche can you explain your conclusion in 
https://github.com/pandas-dev/pandas/pull/24806#issuecomment-455458403 and 
https://github.com/pandas-dev/pandas/pull/24806#issuecomment-455458041 a bit further? I don't see where the precision is being used for parsing (strings or ints), though maybe that's what you're saying?

On 0.23.4

```python
In [3]: pd.DatetimeIndex([2000], dtype='datetime64[s]')
Out[3]: DatetimeIndex(['1970-01-01 00:00:00.000002'], dtype='datetime64[ns]', freq=None)
```

i.e. the same a `datetime64[ns]`. You could imagine it being used as the "unit" parameter for Timestamp for interpreting the integers, but that's not what we did on 0.23.4.

---

So to summarize *my* understanding, which may be wrong

1. In 0.23.4 DatetimeIndex completely ignored the precision.
2. Master currently raises for anything that is non-ns, including no precision.
3. I think we should warn for `dtype='datetime64'` (no precision) since it's ambiguous.
4. I think we should 4raise for `dtype='datetime64[s]'` (non-ns precision) since it's wrong.

those all apply to TimedeltaIndex and Index as well.

**Comment by TomAugspurger** [comment]
Specifically

> Pedantic note: it doesn't really "ignores the precision" in this example case, as it is doing "datetime string parsing", even if you pass dtype=datetime64[ns] it doesn't use the ns precision to interpret the string value '2000'.


Two things there confused me

1. I'm not sure how the precision would affect parsing of strings at all. To me, those don't have an *input* precision (but end up with a precision when they're a Timestamp / array)
2. The `dtype` parameter to Index/DatetimeIndex/TimedeltaIndex means the *desired output dtype*. Not anything to do with parsing, or the input dtype of `data`.

**Comment by jorisvandenbossche** [comment]
Yes, the confusing part (or at least why my comment was also confusing / wrong), is the difference of input vs output dtype. About my comment that the precision is ignored for string parsing, that was indeed assuming that the `dtype` argument would give an indication about the provided input data. But of course, that is not the case. 

Basically, the whole `dtype` argument is just completely useless (apart from consistency with the other constructors, the reason we have it I assume), as there is only one possible dtype (note: this is only true for tz naive of course, but for that there is also the `tz` keyword).

File: pandas/core/arrays/timedeltas.py [hunk 5]
```diff
@@ -997,6 +989,30 @@ def objects_to_td64ns(data, unit="ns", errors="raise"):
...
+                and will raise in a future version. Please pass in
+                'timedelta64[ns]' instead.""")
+            warnings.warn(msg, FutureWarning, stacklevel=4)
+    except TypeError:
+        # extension dtype
+        pass
+
+    try:
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: you can avoid all of this try/except/else if you use ``not is_dtype_equal(dtype, _TD_DTYPE)

File: pandas/core/arrays/timedeltas.py [hunk 6]
```diff
@@ -997,6 +989,30 @@ def objects_to_td64ns(data, unit="ns", errors="raise"):
     return result.view('timedelta64[ns]')
 
 
+def _validate_td64_dtype(dtype):
+    try:
+        if dtype == np.dtype("timedelta64"):
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: same here, use is_dtype_equal

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @TomAugspurger
